### PR TITLE
Support upstream kernel PM netlink events

### PR DESCRIPTION
**Pull request by ossama-othman**
Support the new events in the upstream kernel path management generic netlink API.

Corresponding kernel-side issue: multipath-tcp/mptcp_net-next#60

### CODE REVIEW

File: include/linux/mptcp_upstream.h [hunk 1]
```diff
@@ -36,6 +36,7 @@ enum {
 /* netlink interface */
 #define MPTCP_PM_NAME		"mptcp_pm"
 #define MPTCP_PM_CMD_GRP_NAME	"mptcp_pm_cmds"
+#define MPTCP_PM_EV_GRP_NAME    "mptcp_events"
```
Comment (ossama-othman) [inline include/linux/mptcp_upstream.h:39]: Wasn't this supposed to be `mptcp_pm_events`?

**Review by ossama-othman — COMMENTED** [review]
I'll push the related mptcpd source code changes to this pull request.

File: include/linux/mptcp_upstream.h (hunk 1 above)
Comment (mjmartineau) [reply include/linux/mptcp_upstream.h:39]: Yeah, Matthieu was going to tweak that

Comment (mjmartineau) [reply include/linux/mptcp_upstream.h:39]: Latest export branch tag does have the `mptcp_pm_events` change

Comment (mjmartineau) [reply include/linux/mptcp_upstream.h:39]: Reference: https://github.com/multipath-tcp/mptcp_net-next/commit/1c2dd8ae2f250f7289500962883e568d41378606

Comment (ossama-othman) [reply include/linux/mptcp_upstream.h]: Got it.  Thanks!

Comment (matttbe) [reply include/linux/mptcp_upstream.h]: Thanks to your PR, I just noticed the name was not properly aligned with the others. Just fixed: https://github.com/multipath-tcp/mptcp_net-next/commit/1f2bcf78ff83 (will soon land on the export branch)

**Review by mjmartineau — COMMENTED** [review]
A couple of minor changes to make.

At some point we're going to have to be specific about what kernel versions a given release of mptcpd runs with. For example, current, previous version, and relevant LTS kernels? Oof, I think we're headed toward some build automation to test with a few different kernels.

File: include/mptcpd/private/path_manager.h [hunk 2]
```diff
@@ -46,8 +46,8 @@ struct mptcpd_pm
         /// Mptcpd configuration.
         struct mptcpd_config const *config;
 
-        /// MPTCP path management generic netlink command functions.
-        struct mptcpd_pm_cmd_ops const *cmd_ops;
+        /// Kernel MPTCP generic netlink path manager detailscommand .
```
[the remark this answers is not in the document]
Comment (ossama-othman) [reply include/mptcpd/private/path_manager.h]: Fixed.  s/detailscommand/details/

File: configure.ac [hunk 3]
```diff
@@ -194,7 +194,26 @@ AC_SUBST([ELL_VERSION])
...
+            [Define to 1 if you have the upstream kernel
+            <linux/mptcp.h> header.])
+AH_TEMPLATE([HAVE_LINUX_MPTCP_H_MPTCP_ORG],
+            [Define to 1 if you have the multipath-tcp.org kernel
+            <linux/mptcp.h> header.])
+AC_CHECK_HEADERS([linux/mptcp.h],
+   [
+    AX_CHECK_DEFINE(
+        [linux/mptcp.h],
+        [MPTCP_PM_NAME],
```
[the remark this answers is not in the document]
Comment (ossama-othman) [reply configure.ac]: Thanks Mat!  I pushed your proposed change.

File: src/netlink_pm_upstream.c [hunk 4]
```diff
@@ -596,9 +635,15 @@ static struct mptcpd_pm_cmd_ops const cmd_ops =
         .get_limits  = upstream_get_limits
 };
 
-struct mptcpd_pm_cmd_ops const *mptcpd_get_upstream_cmd_ops(void)
+static struct mptcpd_netlink_pm const npm = {
+        .name    = MPTCP_PM_NAME,
+        .group   = MPTCP_PM_EV_GRP_NAME,
```
[the remark this answers is not in the document]
Comment (ossama-othman) [reply src/netlink_pm_upstream.c:640]: Done.

**Comment by ossama-othman** [comment]
> At some point we're going to have to be specific about what kernel versions a given release of mptcpd runs with. For example, current, previous version, and relevant LTS kernels? Oof, I think we're headed toward some build automation to test with a few different kernels.

That'll be fun. :)   Hopefully the `<linux/mptcp.h>` header won't change much after the new user space genl commands are added.

**Comment by ossama-othman** [comment]
Do we want to support kernels < 5.11?  The `configure.ac` change (ace7f40) to check for `MPTCP_PM_EV_GRP_NAME` instead of `MPTCP_PM_NAME` breaks that support.  For example, I get the following on my Linux box with kernel 5.8 installed:
```
checking for MPTCP_PM_EV_GRP_NAME defined in linux/mptcp.h... no
checking for MPTCP_GENL_EV_GRP_NAME defined in linux/mptcp.h... no
configure: error: Discrepancy in <linux/mptcp.h>.
```
Perhaps it would be better to check for `MPTCP_PM_EV_GRP_NAME` if the upstream kernel `<linux/mptcp.h>` header is detected via `MPTCP_PM_NAME`, and instead use mptcpd's `"mptcpd/private/mptcp_upstream.h"` if the new `MPTCP_PM_EV_GRP_NAME` isn't defined in the system `<linux/mptcp.h>` header.

**Comment by ossama-othman** [comment]
Commit 9d52746 allows the `configure` script run to succeed if an older version of `<linux/mptcp.h>` without the new genl events is detected.

**Review by mjmartineau — APPROVED** [review]
Thanks for the fixes! Looks good.

**Pull request merged**